File: eCaptureFiddler/src/HttpBodyCodec.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Reflection;
using System.Text;

namespace ECaptureFiddler.Core
{
    /// <summary>
    /// Decodes raw HTTP bytes captured by eCapture into plaintext: removes
    /// Transfer-Encoding: chunked framing and recursively decompresses common
    /// Content-Encodings (gzip / deflate(zlib) / br) until the body is no
    /// longer a recognised compression format.
    ///
    /// eCapture text mode forwards raw SSL_read bytes without any decoding, so
    /// gzipped/chunked bodies arrive verbatim and must be decoded here.
    /// </summary>
    public static class HttpBodyCodec
    {
        private static readonly Encoding Latin1 = Encoding.GetEncoding("ISO-8859-1");
        private static readonly byte[] CrlfCrlf = { (byte)'\r', (byte)'\n', (byte)'\r', (byte)'\n' };
        private static readonly byte[] LfLf = { (byte)'\n', (byte)'\n' };

        public static byte[] DecodeHttpMessage(byte[] rawHttp)
        {
            if (rawHttp == null || rawHttp.Length == 0) return rawHttp;
            try
            {
                int sep = IndexOf(rawHttp, CrlfCrlf, 0);
                int sepLen = 4;
                if (sep < 0) { sep = IndexOf(rawHttp, LfLf, 0); sepLen = 2; }
                if (sep < 0) return rawHttp;

                var head = new byte[sep];
                Array.Copy(rawHttp, 0, head, 0, sep);
                int bodyStart = sep + sepLen;
                var body = new byte[rawHttp.Length - bodyStart];
                Array.Copy(rawHttp, bodyStart, body, 0, body.Length);

                string headStr = Latin1.GetString(head);
                string te = HeaderValue(headStr, "Transfer-Encoding");
                string ce = HeaderValue(headStr, "Content-Encoding");

                if (string.IsNullOrEmpty(te) && string.IsNullOrEmpty(ce) && !LooksCompressed(body))
                    return rawHttp;

                byte[] decoded = DecodeBody(body, te, ce);
                string newHead = RewriteHeaders(headStr, decoded.Length);
                byte[] newHeadBytes = Latin1.GetBytes(newHead);

                using (var outMs = new MemoryStream(newHeadBytes.Length + 4 + decoded.Length))
                {
                    outMs.Write(newHeadBytes, 0, newHeadBytes.Length);
                    outMs.WriteByte((byte)'\r'); outMs.WriteByte((byte)'\n');
                    outMs.WriteByte((byte)'\r'); outMs.WriteByte((byte)'\n');
                    outMs.Write(decoded, 0, decoded.Length);
                    return outMs.ToArray();
                }
            }
            catch
            {
                return rawHttp;
            }
        }

        public static byte[] DecodeBody(byte[] body, string transferEncoding, string contentEncoding)
        {
            if (body == null) return Array.Empty<byte>();
            byte[] data = body;

            if (!string.IsNullOrEmpty(transferEncoding) &&
                transferEncoding.ToLowerInvariant().Contains("chunked"))
            {
                data = Dechunk(data);
            }

            if (!string.IsNullOrEmpty(contentEncoding))
            {
                var tokens = contentEncoding.Split(',')
                    .Select(t => t.Trim().ToLowerInvariant())
                    .Where(t => t.Length > 0)
                    .ToList();
                for (int i = tokens.Count - 1; i >= 0; i--)
                    data = ApplyDecode(data, tokens[i]);
            }

            // Magic-byte loop: undeclared / stacked compression.
            for (int round = 0; round < 8; round++)
            {
                if (IsGzip(data))
                {
                    byte[] r = Gunzip(data);
                    if (r.Length == 0 || ByteEquals(r, data)) break;
                    data = r;
                }
                else if (IsZlib(data))
                {
                    byte[] r = Inflate(data, false);
                    if (r.Length == 0 || ByteEquals(r, data)) break;
                    data = r;
                }
                else break;
            }
            return data;
        }

        private static byte[] ApplyDecode(byte[] data, string token)
        {
            switch (token)
            {
                case "gzip":
                case "x-gzip": return Gunzip(data);
                case "deflate": return Inflate(data, false);
                case "br": return Brotli(data);
                default: return data;
            }
        }

        internal static byte[] Dechunk(byte[] input)
        {
            try
            {
                using (var outMs = new MemoryStream())
                {
                    int pos = 0, n = input.Length;
                    while (pos < n && (input[pos] == '\r' || input[pos] == '\n')) pos++;
                    while (pos < n)
                    {
                        int lineEnd = pos;
                        while (lineEnd < n && input[lineEnd] != '\n') lineEnd++;
                        if (lineEnd >= n) break;
                        string sizeLine = Latin1.GetString(input, pos, lineEnd - pos).Trim();
                        int semi = sizeLine.IndexOf(';');
                        if (semi >= 0) sizeLine = sizeLine.Substring(0, semi).Trim();
                        if (sizeLine.Length == 0) { pos = lineEnd + 1; continue; }
                        int chunkSize;
                        try { chunkSize = Convert.ToInt32(sizeLine, 16); }
                        catch { outMs.Write(input, pos, n - pos); break; }
                        pos = lineEnd + 1;
                        if (chunkSize == 0) break;
                        if (pos + chunkSize > n) { outMs.Write(input, pos, n - pos); break; }
                        outMs.Write(input, pos, chunkSize);
                        pos += chunkSize;
                        while (pos < n && (input[pos] == '\r' || input[pos] == '\n')) pos++;
                    }
                    return outMs.ToArray();
                }
            }
            catch { return input; }
        }

        internal static bool IsGzip(byte[] d) =>
            d != null && d.Length >= 3 && d[0] == 0x1F && d[1] == 0x8B && d[2] == 0x08;

        internal static bool IsZlib(byte[] d)
        {
            if (d == null || d.Length < 2) return false;
            int b0 = d[0], b1 = d[1];
            return b0 == 0x78 && ((b0 << 8 | b1) % 31 == 0);
        }

        private static bool LooksCompressed(byte[] d) => IsGzip(d) || IsZlib(d);

        internal static byte[] Gunzip(byte[] input)
        {
            if (input == null || input.Length == 0) return Array.Empty<byte>();
            try
            {
                using (var inMs = new MemoryStream(input))
                using (var gz = new GZipStream(inMs, CompressionMode.Decompress))
                    return ReadAll(gz);
            }
            catch
            {
                // Lenient: eCapture may truncate. Skip 10-byte header, raw inflate.
                if (input.Length <= 10) return input;
                var deflated = new byte[input.Length - 10];
                Array.Copy(input, 10, deflated, 0, deflated.Length);
                byte[] partial = Inflate(deflated, true);
                return partial.Length > 0 ? partial : input;
            }
        }

        internal static byte[] Inflate(byte[] input, bool nowrap)
        {
            if (input == null || input.Length == 0) return Array.Empty<byte>();
            // .NET DeflateStream handles raw deflate (RFC1951). zlib (RFC1950)
            // has a 2-byte header we must skip when nowrap == false.
            byte[] data = input;
            if (!nowrap && input.Length >= 2 && input[0] == 0x78)
            {
                data = new byte[input.Length - 2];
                Array.Copy(input, 2, data, 0, data.Length);
            }
            try
            {
                using (var inMs = new MemoryStream(data))
                using (var df = new DeflateStream(inMs, CompressionMode.Decompress))
                {
                    byte[] r = ReadAll(df);
                    if (r.Length > 0) return r;
                }
            }
            catch { /* fall through */ }

            if (!nowrap && data != input)
            {
                // Retry treating the whole thing as raw deflate.
                try
                {
                    using (var inMs = new MemoryStream(input))
                    using (var df = new DeflateStream(inMs, CompressionMode.Decompress))
                        return ReadAll(df);
                }
                catch { }
            }
            return Array.Empty<byte>();
        }

        // BrotliStream only exists on .NET Core / .NET 5+. Resolve it via
        // reflection so this file still compiles for .NET Framework 4.8 (where
        // brotli is simply unavailable and the input is returned unchanged).
        private static readonly Type BrotliStreamType =
            Type.GetType("System.IO.Compression.BrotliStream, System.IO.Compression.Brotli")
            ?? Type.GetType("System.IO.Compression.BrotliStream, System.IO.Compression");

        internal static byte[] Brotli(byte[] input)
        {
            if (input == null || input.Length == 0) return Array.Empty<byte>();
            if (BrotliStreamType == null) return input; // unsupported on this runtime
            try
            {
                using (var inMs = new MemoryStream(input))
                {
                    var ctor = BrotliStreamType.GetConstructor(new[] { typeof(Stream), typeof(CompressionMode) });
                    using (var br = (Stream)ctor.Invoke(new object[] { inMs, CompressionMode.Decompress }))
                        return ReadAll(br);
                }
            }
            catch { return input; }
        }

        private static byte[] ReadAll(Stream s)
        {
            using (var outMs = new MemoryStream())
            {
                var buf = new byte[8192];
                try
                {
                    int r;
                    while ((r = s.Read(buf, 0, buf.Length)) > 0)
                        outMs.Write(buf, 0, r);
                }
                catch { /* keep partial output on truncated streams */ }
                return outMs.ToArray();
            }
        }

        public static string HeaderValue(string head, string name)
        {
            if (head == null) return "";
            string lname = name.ToLowerInvariant();
            foreach (var line in head.Split('\n'))
            {
                string l = line.TrimEnd('\r');
                int colon = l.IndexOf(':');
                if (colon <= 0) continue;
                if (l.Substring(0, colon).Trim().ToLowerInvariant() == lname)
                    return l.Substring(colon + 1).Trim();
            }
            return "";
        }

        private static string RewriteHeaders(string head, int newLength)
        {
            string[] lines = head.Replace("\r\n", "\n").Split('\n');
            var sb = new StringBuilder();
            bool wroteCl = false;
            for (int i = 0; i < lines.Length; i++)
            {
                string line = lines[i];
                if (i == 0) { sb.Append(line); continue; }
                int colon = line.IndexOf(':');
                string key = colon > 0 ? line.Substring(0, colon).Trim().ToLowerInvariant() : "";
                if (key == "transfer-encoding" || key == "content-encoding") continue;
                if (key == "content-length")
                {
                    sb.Append("\r\n").Append("Content-Length: ").Append(newLength);
                    wroteCl = true;
                    continue;
                }
                if (line.Length == 0) continue;
                sb.Append("\r\n").Append(line);
            }
            if (!wroteCl) sb.Append("\r\n").Append("Content-Length: ").Append(newLength);
            return sb.ToString();
        }

        private static bool ByteEquals(byte[] a, byte[] b)
        {
            if (a == b) return true;
            if (a == null || b == null || a.Length != b.Length) return false;
            for (int i = 0; i < a.Length; i++) if (a[i] != b[i]) return false;
            return true;
        }

        private static int IndexOf(byte[] data, byte[] pattern, int from)
        {
            for (int i = Math.Max(0, from); i <= data.Length - pattern.Length; i++)
            {
                bool ok = true;
                for (int j = 0; j < pattern.Length; j++)
                    if (data[i + j] != pattern[j]) { ok = false; break; }
                if (ok) return i;
            }
            return -1;
        }
    }
}

File: eCaptureBurp/src/main/java/com/ecapture/burp/export/HttpBodyCodec.java
package com.ecapture.burp.export;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.GZIPInputStream;
import java.util.zip.Inflater;

/**
 * Decodes raw HTTP message bytes captured by eCapture so they display as
 * plaintext: removes Transfer-Encoding: chunked framing and recursively
 * decompresses common Content-Encodings (gzip / deflate(zlib) / br) until the
 * body is no longer a recognised compression format.
 *
 * eCapture's text mode forwards the raw SSL_read bytes without any decoding, so
 * a gzipped/chunked body arrives verbatim. Burp does not always auto-decode it
 * (e.g. when the body is chunked, fragmented, or the encoding is undeclared),
 * which is why we decode it ourselves here.
 */
public final class HttpBodyCodec {

    private HttpBodyCodec() {}

    private static final byte[] CRLFCRLF = {'\r', '\n', '\r', '\n'};
    private static final byte[] LFLF = {'\n', '\n'};

    /**
     * Decode a full HTTP message (request or response): de-chunk + decompress
     * the body, then fix up the headers (drop Transfer-Encoding /
     * Content-Encoding, set Content-Length to the decoded length).
     * Returns the original bytes unchanged if no header terminator is found or
     * on any error.
     */
    public static byte[] decodeHttpMessage(byte[] rawHttp) {
        if (rawHttp == null || rawHttp.length == 0) {
            return rawHttp;
        }
        try {
            int sep = indexOf(rawHttp, CRLFCRLF, 0);
            int sepLen = 4;
            if (sep < 0) {
                sep = indexOf(rawHttp, LFLF, 0);
                sepLen = 2;
            }
            if (sep < 0) {
                // No header/body boundary: nothing to decode.
                return rawHttp;
            }

            byte[] head = new byte[sep];
            System.arraycopy(rawHttp, 0, head, 0, sep);
            int bodyStart = sep + sepLen;
            byte[] body = new byte[rawHttp.length - bodyStart];
            System.arraycopy(rawHttp, bodyStart, body, 0, body.length);

            String headStr = new String(head, StandardCharsets.ISO_8859_1);
            String transferEncoding = headerValue(headStr, "Transfer-Encoding");
            String contentEncoding = headerValue(headStr, "Content-Encoding");

            if ((transferEncoding == null || transferEncoding.isEmpty())
                    && (contentEncoding == null || contentEncoding.isEmpty())
                    && !looksCompressed(body)) {
                // Nothing to do.
                return rawHttp;
            }

            byte[] decoded = decodeBody(body, transferEncoding, contentEncoding);

            String newHead = rewriteHeaders(headStr, decoded.length);
            byte[] newHeadBytes = newHead.getBytes(StandardCharsets.ISO_8859_1);

            ByteArrayOutputStream out = new ByteArrayOutputStream(newHeadBytes.length + 4 + decoded.length);
            out.write(newHeadBytes);
            out.write('\r');
            out.write('\n');
            out.write('\r');
            out.write('\n');
            out.write(decoded);
            return out.toByteArray();
        } catch (Exception e) {
            return rawHttp;
        }
    }

    /**
     * Decode just a body: de-chunk (if chunked) then recursively decompress.
     */
    public static byte[] decodeBody(byte[] body, String transferEncoding, String contentEncoding) {
        if (body == null) {
            return new byte[0];
        }
        byte[] data = body;

        // 1. De-chunk if Transfer-Encoding indicates chunked framing.
        if (transferEncoding != null && transferEncoding.toLowerCase().contains("chunked")) {
            data = dechunk(data);
        }

        // 2. Apply declared Content-Encoding tokens in reverse order.
        //    "Content-Encoding: gzip, deflate" means deflate was applied last,
        //    so we must inflate then gunzip (reverse of application order).
        if (contentEncoding != null && !contentEncoding.isEmpty()) {
            List<String> tokens = new ArrayList<>();
            for (String t : contentEncoding.split(",")) {
                String tok = t.trim().toLowerCase();
                if (!tok.isEmpty()) {
                    tokens.add(tok);
                }
            }
            for (int i = tokens.size() - 1; i >= 0; i--) {
                data = applyDecode(data, tokens.get(i));
            }
        }

        // 3. Magic-byte loop: catch undeclared / stacked compression. Keep
        //    decompressing while the data still starts with a known magic.
        for (int round = 0; round < 8; round++) {
            if (isGzip(data)) {
                byte[] r = gunzip(data);
                if (r.length == 0 || java.util.Arrays.equals(r, data)) break;
                data = r;
            } else if (isZlib(data)) {
                byte[] r = inflate(data, false);
                if (r.length == 0 || java.util.Arrays.equals(r, data)) break;
                data = r;
            } else {
                break;
            }
        }
        return data;
    }

    private static byte[] applyDecode(byte[] data, String token) {
        switch (token) {
            case "gzip":
            case "x-gzip":
                return gunzip(data);
            case "deflate":
                return inflate(data, false);
            case "br":
                return brotli(data);
            case "identity":
            case "":
                return data;
            default:
                return data;
        }
    }

    /** Parse Transfer-Encoding: chunked framing into the raw body. */
    static byte[] dechunk(byte[] in) {
        try {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            int pos = 0;
            int n = in.length;
            // Tolerate leading CRLF/whitespace before the first chunk-size.
            while (pos < n && (in[pos] == '\r' || in[pos] == '\n')) {
                pos++;
            }
            while (pos < n) {
                int lineEnd = pos;
                while (lineEnd < n && in[lineEnd] != '\n') {
                    lineEnd++;
                }
                if (lineEnd >= n) break;
                String sizeLine = new String(in, pos, lineEnd - pos, StandardCharsets.ISO_8859_1).trim();
                // chunk-size may have extensions: "1a;name=value"
                int semi = sizeLine.indexOf(';');
                if (semi >= 0) {
                    sizeLine = sizeLine.substring(0, semi).trim();
                }
                if (sizeLine.isEmpty()) {
                    pos = lineEnd + 1;
                    continue;
                }
                int chunkSize;
                try {
                    chunkSize = Integer.parseInt(sizeLine, 16);
                } catch (NumberFormatException e) {
                    // Not chunked after all - return what we have plus the rest.
                    out.write(in, pos, n - pos);
                    break;
                }
                pos = lineEnd + 1; // move past the size line
                if (chunkSize == 0) {
                    break; // last chunk
                }
                if (pos + chunkSize > n) {
                    // Truncated chunk: keep whatever data remains.
                    out.write(in, pos, n - pos);
                    break;
                }
                out.write(in, pos, chunkSize);
                pos += chunkSize;
                // Skip trailing CRLF after chunk data.
                while (pos < n && (in[pos] == '\r' || in[pos] == '\n')) {
                    pos++;
                }
            }
            return out.toByteArray();
        } catch (Exception e) {
            return in;
        }
    }

    static boolean isGzip(byte[] d) {
        return d != null && d.length >= 3
                && (d[0] & 0xFF) == 0x1F && (d[1] & 0xFF) == 0x8B && (d[2] & 0xFF) == 0x08;
    }

    static boolean isZlib(byte[] d) {
        if (d == null || d.length < 2) return false;
        int b0 = d[0] & 0xFF;
        int b1 = d[1] & 0xFF;
        // zlib: CMF=0x78 (deflate, 32K window) and (CMF*256+FLG) % 31 == 0
        return b0 == 0x78 && ((b0 << 8 | b1) % 31 == 0);
    }

    private static boolean looksCompressed(byte[] d) {
        return isGzip(d) || isZlib(d);
    }

    static byte[] gunzip(byte[] in) {
        if (in == null || in.length == 0) return new byte[0];
        try (GZIPInputStream gz = new GZIPInputStream(new ByteArrayInputStream(in))) {
            return readAll(gz);
        } catch (Exception e) {
            // Lenient: eCapture may truncate; return whatever inflate recovers.
            byte[] partial = gunzipRaw(in);
            return partial.length > 0 ? partial : in;
        }
    }

    /** Raw inflate of gzip payload skipping the 10-byte header (best-effort). */
    private static byte[] gunzipRaw(byte[] in) {
        try {
            if (in.length <= 10) return new byte[0];
            byte[] deflated = new byte[in.length - 10];
            System.arraycopy(in, 10, deflated, 0, deflated.length);
            return inflate(deflated, true);
        } catch (Exception e) {
            return new byte[0];
        }
    }

    static byte[] inflate(byte[] in, boolean nowrap) {
        if (in == null || in.length == 0) return new byte[0];
        Inflater inflater = new Inflater(nowrap);
        try {
            inflater.setInput(in);
            ByteArrayOutputStream out = new ByteArrayOutputStream(Math.max(64, in.length * 3));
            byte[] buf = new byte[8192];
            while (!inflater.finished()) {
                int count = inflater.inflate(buf);
                if (count == 0) {
                    if (inflater.needsInput() || inflater.needsDictionary()) {
                        break;
                    }
                }
                out.write(buf, 0, count);
            }
            byte[] result = out.toByteArray();
            if (result.length == 0 && !nowrap) {
                // Some servers send raw deflate without zlib header.
                return inflate(in, true);
            }
            return result;
        } catch (Exception e) {
            if (!nowrap) {
                return inflate(in, true);
            }
            return new byte[0];
        } finally {
            inflater.end();
        }
    }

    static byte[] brotli(byte[] in) {
        if (in == null || in.length == 0) return new byte[0];
        try {
            org.brotli.dec.BrotliInputStream bis =
                    new org.brotli.dec.BrotliInputStream(new ByteArrayInputStream(in));
            try {
                return readAll(bis);
            } finally {
                bis.close();
            }
        } catch (Throwable e) {
            return in;
        }
    }

    private static byte[] readAll(java.io.InputStream is) throws Exception {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        byte[] buf = new byte[8192];
        int r;
        try {
            while ((r = is.read(buf)) != -1) {
                out.write(buf, 0, r);
            }
        } catch (Exception e) {
            // Keep partial output on truncated streams.
        }
        return out.toByteArray();
    }

    /**
     * Return the (comma-joined) value of an HTTP header, case-insensitive.
     */
    public static String headerValue(String head, String name) {
        if (head == null) return "";
        String lname = name.toLowerCase();
        for (String line : head.split("\r\n|\n")) {
            int colon = line.indexOf(':');
            if (colon <= 0) continue;
            if (line.substring(0, colon).trim().toLowerCase().equals(lname)) {
                return line.substring(colon + 1).trim();
            }
        }
        return "";
    }

    /**
     * Drop Transfer-Encoding and Content-Encoding headers and set
     * Content-Length to the decoded length.
     */
    private static String rewriteHeaders(String head, int newLength) {
        String[] lines = head.split("\r\n|\n", -1);
        StringBuilder sb = new StringBuilder();
        boolean wroteContentLength = false;
        for (int i = 0; i < lines.length; i++) {
            String line = lines[i];
            if (i == 0) {
                sb.append(line); // status / request line
                continue;
            }
            int colon = line.indexOf(':');
            String key = colon > 0 ? line.substring(0, colon).trim().toLowerCase() : "";
            if (key.equals("transfer-encoding") || key.equals("content-encoding")) {
                continue; // strip
            }
            if (key.equals("content-length")) {
                sb.append("\r\n").append("Content-Length: ").append(newLength);
                wroteContentLength = true;
                continue;
            }
            if (line.isEmpty()) {
                continue;
            }
            sb.append("\r\n").append(line);
        }
        if (!wroteContentLength) {
            sb.append("\r\n").append("Content-Length: ").append(newLength);
        }
        return sb.toString();
    }

    private static int indexOf(byte[] data, byte[] pattern, int from) {
        outer:
        for (int i = Math.max(0, from); i <= data.length - pattern.length; i++) {
            for (int j = 0; j < pattern.length; j++) {
                if (data[i + j] != pattern[j]) {
                    continue outer;
                }
            }
            return i;
        }
        return -1;
    }
}
